redaxo/src/addons/media_manager/lib/effects/effect_img2img.php

```diff
@@ -0,0 +1,157 @@
+            'ext' => 'gif',
+            'content-type' => 'image/gif',
+        ],
+        'webp' => [
+            'ext' => 'webp',
+            'content-type' => 'image/webp',
+        ],
+    ];
+
+    private static $convert_tos = ['jpg', 'png', 'gif', 'webp'];
+    private static $convert_to_default = 'jpg';
+    private static $convert_to_default_quality = 80;
+
+    public function execute()
+    {
+        $this->media->asImage();
+        $imageObject = $this->media->getImage();
+        if (null === $imageObject) {
+            return;
+        }
+
+        if (!isset(self::$convert_to[$this->params['convert_to']])) {
+            $convert_to = self::$convert_to[self::$convert_to_default];
+        } else {
+            $convert_to = self::$convert_to[$this->params['convert_to']];
+        }
+
+        $quality = 0;
+        switch ($convert_to['ext']) {
+            case 'jpg':
+                $quality = rex_addon::get('media_manager')->getConfig('jpg_quality');
+                break;
+
+            case 'webp':
+                $quality = rex_addon::get('media_manager')->getConfig('webp_quality');
+                break;
+
+            case 'png':
+                $quality = rex_addon::get('media_manager')->getConfig('png_compression');
+                break;
+        }
+
+        if (!$quality) {
+            $quality = self::$convert_to_default_quality;
+        }
+        if ($this->params['quality'] && (int) $this->params['quality'] > 0) {
+            $quality = $this->params['quality'];
+        }
+        if (100 < (int) $quality) {
+            $quality = 100;
+        }
+        if ('png' == $convert_to['ext']) {
+            if (9 < (int) $quality) {
+                $quality = 9;
+            }
+        }
+
+        $from_path = realpath($this->media->getMediaPath());
+        $ext = rex_file::extension($from_path);
+
+        // skip if extension is not in list
+        if (!in_array(strtolower($ext), self::$convert_types)) {
+            return;
+        }
+
+        $filename = $this->media->getMediaFilename();
+        $filename_wo_ext = substr($filename, 0, (strlen($filename) - strlen($ext)));
+        $targetFilename = $filename_wo_ext . $convert_to['ext'];
+        $to_path = rex_path::addonCache('media_manager', 'media_manager__convert_img2img_'.md5($this->media->getMediaPath()).'_'.$targetFilename);
+
+        // save
+        $funcOut = 'image' . strtolower('jpg' == $this->params['convert_to'] ? 'jpeg' : $this->params['convert_to']);
+
+        // no support for the output function
+        if (!function_exists($funcOut)) {
+            rex_file::delete($to_path);
+            return;
+        }
+
+        imagepalettetotruecolor($imageObject);
+        $saved = @call_user_func($funcOut, $imageObject, $to_path, $quality);
+
+        if (!$saved) {
+            return;
+        }
+
+        $funcIn = 'imagecreatefrom' . strtolower('jpg' == $this->params['convert_to'] ? 'jpeg' : $this->params['convert_to']);
```

Comment:
Es ist eigentlich unnötig, dass hier das Bild per gd erneut eingelesen wird. Es sollte reichen, den sourcePath auf die Temp-Datei zu setzen, und halt Format etc, ähnlich [wie in convert2img](https://github.com/redaxo/redaxo/blob/11da0bc97518ab4cb1d298df158d117e03788dc4/redaxo/src/addons/media_manager/lib/effects/effect_convert2img.php#L113-L117).
In `setSourcePath` wird die alte gd-resource automatisch entfernt. Und neu aufgemacht wird sie dann nur, wenn ein späterer Effekt erneut `asImage` aufruft.

Reply:
Ich werde noch mal komplett drübergehen über den Code und optimieren @gharlan